src/code/speed_meter.c

```diff
@@ -1,18 +1,266 @@
+        width = ((f64)*sSpeedMeterTimeEntryPtr->time / gIrqMgrRetraceTime) * 64.0;
+        sSpeedMeterTimeEntryPtr->x = baseX + width;
+        sSpeedMeterTimeEntryPtr++;
+    }
+
+    View_Init(&view, gfxCtx);
+    view.flags = VIEW_VIEWPORT | VIEW_PROJECTION_ORTHO;
+
+    SET_FULLSCREEN_VIEWPORT_DYNAMIC(&view);
+
+    gfx = OVERLAY_DISP;
+    View_ApplyTo(&view, &gfx);
+
+    gDPPipeSync(gfx++);
+    gDPSetOtherMode(gfx++,
+                    G_AD_PATTERN | G_CD_MAGICSQ | G_CK_NONE | G_TC_CONV | G_TF_POINT | G_TT_NONE | G_TL_TILE |
+                        G_TD_CLAMP | G_TP_NONE | G_CYC_FILL | G_PM_NPRIMITIVE,
+                    G_AC_NONE | G_ZS_PIXEL | G_RM_NOOP | G_RM_NOOP2);
+
+    gDrawRect(gfx++, GPACK_RGBA5551(0, 0, 255, 1), baseX + 64 * 0, uly, baseX + 64 * 1, lry);
+    gDrawRect(gfx++, GPACK_RGBA5551(0, 255, 0, 1), baseX + 64 * 1, uly, baseX + 64 * 2, lry);
+    gDrawRect(gfx++, GPACK_RGBA5551(255, 0, 0, 1), baseX + 64 * 2, uly, baseX + 64 * 3, lry);
+    gDrawRect(gfx++, GPACK_RGBA5551(255, 0, 255, 1), baseX + 64 * 3, uly, baseX + 64 * 4, lry);
+
+    sSpeedMeterTimeEntryPtr = &sSpeedMeterTimeEntryArray[0];
+    for (i = 0; i < ARRAY_COUNT(sSpeedMeterTimeEntryArray); i++) {
+        gDrawRect(gfx++, sSpeedMeterTimeEntryPtr->color, baseX, lry + sSpeedMeterTimeEntryPtr->y,
+                  sSpeedMeterTimeEntryPtr->x, lry + sSpeedMeterTimeEntryPtr->y + 1);
+        sSpeedMeterTimeEntryPtr++;
+    }
+    gDPPipeSync(gfx++);
+
+    OVERLAY_DISP = gfx;
+
+    CLOSE_DISPS(gfxCtx);
+}
+
+void SpeedMeter_InitAllocEntry(SpeedMeterAllocEntry* this, u32 maxVal, u32 val, u16 backColor, u16 foreColor, u32 ulx,
+                               u32 lrx, u32 uly, u32 lry) {
+    this->maxVal = maxVal;
+    this->val = val;
+    this->backColor = backColor;
+    this->foreColor = foreColor;
+    this->ulx = ulx;
+    this->lrx = lrx;
+    this->uly = uly;
+    this->lry = lry;
+}
+
+void SpeedMeter_DrawAllocEntry(SpeedMeterAllocEntry* this, GraphicsContext* gfxCtx) {
+    s32 usedOff;
+    View view;
+    Gfx* gfx;
+
+    if (this->maxVal != 0) {
+        OPEN_DISPS(gfxCtx);
+
+        View_Init(&view, gfxCtx);
+        view.flags = VIEW_VIEWPORT | VIEW_PROJECTION_ORTHO;
+
+        SET_FULLSCREEN_VIEWPORT_DYNAMIC(&view);
+
+        gfx = OVERLAY_DISP;
+        View_ApplyTo(&view, &gfx);
+
+        gDPPipeSync(gfx++);
+        gDPSetOtherMode(gfx++,
+                        G_AD_PATTERN | G_CD_MAGICSQ | G_CK_NONE | G_TC_CONV | G_TF_POINT | G_TT_NONE | G_TL_TILE |
+                            G_TD_CLAMP | G_TP_NONE | G_CYC_FILL | G_PM_NPRIMITIVE,
+                        G_AC_NONE | G_ZS_PIXEL | G_RM_NOOP | G_RM_NOOP2);
+
+        usedOff = ((this->lrx - this->ulx) * this->val) / this->maxVal + this->ulx;
+        gDrawRect(gfx++, this->backColor, usedOff, this->uly, this->lrx, this->lry);
+        gDrawRect(gfx++, this->foreColor, this->ulx, this->uly, usedOff, this->lry);
+
+        gDPPipeSync(gfx++);
+
+        //! FAKE:
```

Comment:
Do you have a decomp.me for this?

Reply:
https://decomp.me/scratch/mstcw

Reply:
Ah, sorry, I should've clarified. I try to put decomp.me links below `//! FAKE` comments so people can attempt to fix them. However, in this case, it's not super necessary, since you could just delete the following line and try it for yourself. So I'll leave it up to you whether you want to put that in or not.

Reply:
Ah, tbh working locally with this is probably best anyways, due to the graphic macros.

Reply:
Yeah, fair enough. No need to change anything then; my approval stands with the code as-is.